net/net_test.go

```diff
@@ -236,3 +237,61 @@ func TestIPNetsDoNotContain(t *testing.T) {
 		})
 	}
 }
+
+func TestSchemeHost(t *testing.T) {
+	for _, tt := range []struct {
+		input  string
+		scheme string
+		host   string
+		err    error
+	}{
+		{"http://example.com", "http", "example.com:80", nil},
+		{"http://example.com:80", "http", "example.com:80", nil},
+		{"http://example.com:8080", "http", "example.com:8080", nil},
+
+		{"https://example.com", "https", "example.com:443", nil},
+		{"https://example.com:443", "https", "example.com:443", nil},
+		{"https://example.com:8080", "https", "example.com:8080", nil},
+
+		{"postgres://example.com", "", "", errors.New(`parse "postgres://example.com": empty port`)},
+		{"postgres://example.com:5432", "postgres", "example.com:5432", nil},
+		{"postgresql://example.com", "", "", errors.New(`parse "postgresql://example.com": empty port`)},
+		{"postgresql://example.com:5432", "postgresql", "example.com:5432", nil},
+
+		{"someprotocol://example.com", "", "", errors.New(`parse "someprotocol://example.com": empty port`)},
+		{"someprotocol://example.com:12345", "someprotocol", "example.com:12345", nil},
+
+		{"example.com", "", "", errors.New(`parse "example.com": invalid URI for request`)},
+		{"example.com/", "", "", errors.New(`parse "example.com/": invalid URI for request`)},
+		{"example.com:80", "", "", errors.New(`parse "example.com:80": empty host`)},
+
+		{"hTTp://exAMPLe.com", "http", "example.com:80", nil},
+
+		{"http://example.com/foo/bar", "http", "example.com:80", nil},
+		{"http://example.com:80/foo/bar", "http", "example.com:80", nil},
+		{"http://example.com:8080/foo/bar", "http", "example.com:8080", nil},
+
+		{"http://example.com?foo=bar", "http", "example.com:80", nil},
+		{"http://example.com:80?foo=bar", "http", "example.com:80", nil},
+		{"http://example.com:8080?foo=bar", "http", "example.com:8080", nil},
+	} {
+		t.Run(tt.input, func(t *testing.T) {
+			scheme, host, err := SchemeHost(tt.input)
+			if tt.err == nil && err != nil {
+				t.Errorf("unexpected error: '%v'", err)
+			}
+			if tt.err != nil && err == nil {
+				t.Errorf("failed to fail")
+			}
+			if tt.err != nil && err != nil && tt.err.Error() != err.Error() {
+				t.Errorf("unexpected error: '%v'", err)
+			}
```

Comment:
This would be so much cleaner with assert:
```go
if tc.err != "" {
     assert.EqualError(t, err, tc.err)
} else {
     if assert.NoError(t, err) {
          assert.Equal(t, tc.scheme, scheme)
          assert.Equal(t, tc.host, host)
     }
}
```